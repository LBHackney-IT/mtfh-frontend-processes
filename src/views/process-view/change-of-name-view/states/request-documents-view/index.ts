export * from "./request-documents-view";
