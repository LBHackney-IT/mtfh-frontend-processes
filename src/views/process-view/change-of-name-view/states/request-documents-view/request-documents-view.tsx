import { CheckSupportingDocuments } from "../../../../../components/check-supporting-documents";
import { locale } from "../../../../../services";
import { ProcessComponentProps } from "../../../../../types";
import { BulletWithExplanation } from "../../../shared/process-components";
import { changeOfNameDocuments } from "../../view-utils";

import { Center, Heading, List, Spinner, Text } from "@mtfh/common/lib/components";
import { useErrorCodes } from "@mtfh/common/lib/hooks";

export const RequestDocumentsView = ({
  process,
  mutate,
  optional,
}: ProcessComponentProps): JSX.Element => {
  const errorMessages = useErrorCodes();

  if (!errorMessages) {
    return (
      <Center>
        <Spinner />
      </Center>
    );
  }

  return !optional.closeProcessReason ? (
    <div data-testid="changeofname-RequestDocuments">
      <Heading variant="h3">{locale.supportingDocuments}</Heading>

      <Text size="sm">
        The tenant needs to provide documents as proof to support their application:
      </Text>
      <List variant="bullets">
        <BulletWithExplanation
          text="Two forms of proof of identity. At least one must be photographic ID. Both must contain the tenants new name"
          explanation="for example: valid passport, driving licence, bank statement, utility bill"
        />
        <div>
          <Text size="sm" style={{ fontWeight: "bold" }}>
            One of the following documents containing the tenantâ€™s new name:
          </Text>
          <ol
            className="lbh-body-s"
            type="1"
            style={{ marginLeft: "2em", marginTop: 15, fontWeight: "bold" }}
          >
            {changeOfNameDocuments.map((item) => (
              <li key={item} style={{ marginTop: 0 }}>
                {item}
              </li>
            ))}
          </ol>
        </div>
      </List>

      <CheckSupportingDocuments process={process} mutate={mutate} optional={optional} />
    </div>
  ) : (
    <></>
  );
};
