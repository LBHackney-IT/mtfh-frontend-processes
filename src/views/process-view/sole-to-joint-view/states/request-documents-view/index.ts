export * from "./request-dcouments-view";
