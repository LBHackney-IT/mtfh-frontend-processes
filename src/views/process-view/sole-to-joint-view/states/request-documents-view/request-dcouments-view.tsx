import { CheckSupportingDocuments } from "../../../../../components/check-supporting-documents";
import { locale } from "../../../../../services";
import { IProcess, ProcessComponentProps } from "../../../../../types";
import { BulletWithExplanation } from "../../../shared/process-components";
import { EligibilityChecksPassedBox } from "../shared";

import { useTenure } from "@mtfh/common/lib/api/tenure/v1";
import {
  Center,
  ErrorSummary,
  Heading,
  List,
  Spinner,
  Text,
} from "@mtfh/common/lib/components";
import { useErrorCodes } from "@mtfh/common/lib/hooks";

export interface RequestDocumentsViewProps extends ProcessComponentProps {
  processConfig: IProcess;
}

export const RequestDocumentsView = ({
  process,
  mutate,
  optional,
}: RequestDocumentsViewProps) => {
  const { closeProcessReason } = optional;
  const errorMessages = useErrorCodes();
  const { data: tenure, error } = useTenure(process.targetId);

  if (error) {
    return (
      <ErrorSummary
        id="request-documents-view"
        title={locale.errors.unableToFetchRecord}
        description={locale.errors.unableToFetchRecordDescription}
      />
    );
  }

  if (!tenure || !errorMessages) {
    return (
      <Center>
        <Spinner />
      </Center>
    );
  }

  const tenant = tenure?.householdMembers.find((m) => m.isResponsible);

  return (
    <div data-testid="soletojoint-RequestDocuments">
      <EligibilityChecksPassedBox />
      {!closeProcessReason && (
        <>
          <Heading variant="h3">{locale.supportingDocuments}</Heading>
          <Text>
            The following documentation is required from the secure tenant and/or proposed
            tenant, as proof to support their application:
          </Text>
          <List variant="bullets">
            <BulletWithExplanation
              text="Secure and Proposed tenant: Two forms of proof of identity for both the
              secure and proposed tenant. At least one each must be photographic ID"
              explanation="for example: valid passport, driving licence, bank statement, utility bill"
            />
            <BulletWithExplanation
              text="Proposed tenant: Proof of immigration status"
              explanation="for example: passport, home office letter, embassy letter, immigration status document"
            />
            <BulletWithExplanation
              text="Proposed tenant: Proof of relationship to the existing tenant"
              explanation="for example: marriage or civil partner certificate"
            />
            <BulletWithExplanation
              text="Proposed tenant: Proof of co-habitation: three documents proving 12 months
                residency at the property. If marriage certificate provided, any Proof of
                Address can be accepted."
              explanation="for example: letter, utility bill, council tax bill"
            />
          </List>

          <CheckSupportingDocuments
            process={process}
            mutate={mutate}
            optional={{ tenant, declaration: true }}
          />
        </>
      )}
    </div>
  );
};
