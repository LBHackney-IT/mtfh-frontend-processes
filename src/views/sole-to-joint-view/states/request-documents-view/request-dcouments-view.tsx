import { Form, Formik } from "formik";

import { SoleToJointHeader } from "../../../../components";
import { DateTimeFields } from "../../../../components/appointment-form/appointment-form";
import {
  RequestDocumentsFormData,
  requestDocumentsSchema,
} from "../../../../schemas/request-documents";
import { locale } from "../../../../services";
import { IProcess } from "../../../../types";
import { dateToString, stringToDate } from "../../../../utils/date";
import {
  BulletWithExplanation,
  EligibilityChecksPassedBox,
  TenantContactDetails,
} from "../shared";

import { Process, editProcess } from "@mtfh/common/lib/api/process/v1";
import { useTenure } from "@mtfh/common/lib/api/tenure/v1";
import {
  Button,
  Center,
  ErrorSummary,
  Heading,
  InlineField,
  List,
  Radio,
  RadioGroup,
  Spinner,
  Text,
} from "@mtfh/common/lib/components";
import { useErrorCodes } from "@mtfh/common/lib/hooks";

export interface RequestDocumentsViewProps {
  process: Process;
  processConfig: IProcess;
  mutate: any;
}

export const RequestDocumentsView = ({
  process,
  processConfig,
  mutate,
}: RequestDocumentsViewProps) => {
  const errorMessages = useErrorCodes();
  const { data: tenure, error } = useTenure(process.targetId);

  if (error) {
    return (
      <ErrorSummary
        id="request-documents-view"
        title={locale.errors.unableToFetchRecord}
        description={locale.errors.unableToFetchRecordDescription}
      />
    );
  }

  if (!tenure || !errorMessages) {
    return (
      <Center>
        <Spinner />
      </Center>
    );
  }

  const stateConfig = processConfig.states.breachChecksPassed;
  const tenant = tenure?.householdMembers.find((m) => m.isResponsible);

  return (
    <div data-testid="soletojoint-RequestDocuments">
      <SoleToJointHeader processConfig={processConfig} process={process} />
      <EligibilityChecksPassedBox />
      <Heading variant="h3">{locale.supportingDocuments}</Heading>
      <Text size="sm">
        The following documentation is required from the secure tenant and/or proposed
        tenant, as proof to support their application:
      </Text>
      <List variant="bullets">
        <BulletWithExplanation
          text="Secure and Proposed tenant: Two forms of proof of identity for both the
              secure and proposed tenant. At least one each must be photographic ID"
          explanation="for example: valid passport, driving licence, bank statement, utility bill"
        />
        <BulletWithExplanation
          text="Proposed tenant: Proof of immigration status"
          explanation="for example: passport, home office letter, embassy letter, immigration status document"
        />
        <BulletWithExplanation
          text="Proposed tenant: Proof of relationship to the existing tenant"
          explanation="for example: marriage or civil partner certificate"
        />
        <BulletWithExplanation
          text="Proposed tenant: Proof of co-habitation: three documents proving 12 months
                residency at the property. If marriage certificate provided, any Proof of
                Address can be accepted."
          explanation="for example: letter, utility bill, council tax bill"
        />
      </List>
      <Heading variant="h3">Checking supporting documents</Heading>
      <Text size="sm">
        You can request supporting documents through the Document Evidence Store or you
        can make an appointment with the tenant to check supporting documents in-person.
      </Text>
      {tenant ? <TenantContactDetails tenant={tenant} /> : <Text>Tenant not found.</Text>}
      <Formik<RequestDocumentsFormData>
        initialValues={{
          requestType: "",
          day: "",
          month: "",
          year: "",
          hour: "",
          minute: "",
          amPm: "",
        }}
        validateOnBlur={false}
        validateOnChange={false}
        validationSchema={requestDocumentsSchema(errorMessages)}
        onSubmit={async (values) => {
          const { requestType, day, month, year, hour, minute, amPm } = values;
          let processTrigger = stateConfig.triggers.requestDocumentsDes;
          let formData = {};
          if (requestType === "manual") {
            processTrigger = stateConfig.triggers.requestDocumentsAppointment;
            const date = stringToDate(
              `${day}/${month}/${year} ${hour}:${minute} ${amPm.toUpperCase()}`,
              "dd/MM/yyyy hh:mm a",
            );
            formData = {
              appointmentDateTime: dateToString(date, "yyyy-MM-dd'T'HH:mm:ss"),
            };
          }
          try {
            await editProcess({
              id: process.id,
              processName: process?.processName,
              etag: process.etag || "",
              processTrigger,
              formData,
              documents: [],
            });
            mutate();
          } catch (e: any) {
            console.log(e.response?.status || 500);
          }
        }}
      >
        {(props) => {
          const {
            values,
            values: { requestType },
            setFieldValue,
          } = props;
          return (
            <Form
              noValidate
              id="request-documents-form"
              className="request-documents-form"
            >
              <RadioGroup>
                <InlineField name="requestType" type="radio">
                  <Radio
                    id="requestType-automatic"
                    value="automatic"
                    onClick={() => {
                      if (requestType === "manual") {
                        setFieldValue("day", "01");
                        setFieldValue("month", "01");
                        setFieldValue("year", "3000");
                        setFieldValue("hour", "01");
                        setFieldValue("minute", "01");
                        setFieldValue("amPm", "am");
                      }
                    }}
                  >
                    Request documents electronically
                  </Radio>
                </InlineField>
                <InlineField name="requestType" type="radio">
                  <Radio
                    id="requestType-manual"
                    value="manual"
                    onClick={() => {
                      if (requestType === "automatic") {
                        setFieldValue("day", "");
                        setFieldValue("month", "");
                        setFieldValue("year", "");
                        setFieldValue("hour", "");
                        setFieldValue("minute", "");
                        setFieldValue("amPm", "");
                      }
                    }}
                  >
                    I have made an appointment to check supporting documents
                  </Radio>
                </InlineField>
              </RadioGroup>
              {requestType === "manual" && <DateTimeFields />}
              <Button
                type="submit"
                disabled={
                  !Object.values(values).some((value) => {
                    return value !== "";
                  })
                }
              >
                Next
              </Button>
            </Form>
          );
        }}
      </Formik>
    </div>
  );
};
