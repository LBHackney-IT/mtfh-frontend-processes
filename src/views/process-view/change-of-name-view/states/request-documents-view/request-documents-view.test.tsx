import React from "react";

import {
  getReferenceDataV1,
  mockContactDetailsV2,
  mockPersonV1,
  mockProcessV1,
  render,
  server,
} from "@hackney/mtfh-test-utils";
import { screen, waitForElementToBeRemoved } from "@testing-library/react";

import { RequestDocumentsView } from "./request-documents-view";

import * as contactDetailsService from "@mtfh/common/lib/api/contact-details/v2/service";
import { ContactDetailsResponse } from "@mtfh/common/lib/api/contact-details/v2/service";
import { AxiosSWRResponse } from "@mtfh/common/lib/http";

const options = {
  url: "/processes/changeofname/e63e68c7-84b0-3a48-b450-896e2c3d7735",
  path: "/processes/:processName/:processId",
};

describe("changeofname/request-documents-view", () => {
  beforeEach(() => {
    jest.resetModules();
    server.use(getReferenceDataV1());
  });

  test("it renders RequestDocuments correctly on state NameSubmitted", async () => {
    jest.spyOn(contactDetailsService, "useContactDetails").mockReturnValue({
      data: mockContactDetailsV2,
    } as AxiosSWRResponse<ContactDetailsResponse>);
    const { container } = render(
      <RequestDocumentsView
        process={{
          ...mockProcessV1,
          currentState: { ...mockProcessV1.currentState, state: "NameSubmitted" },
        }}
        mutate={() => {}}
        optional={{
          person: mockPersonV1,
        }}
      />,
      options,
    );
    await waitForElementToBeRemoved(screen.queryAllByText(/Loading/));
    expect(container).toMatchSnapshot();
  });
});
