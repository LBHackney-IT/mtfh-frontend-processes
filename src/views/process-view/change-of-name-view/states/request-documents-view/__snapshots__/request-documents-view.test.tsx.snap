// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`changeofname/request-documents-view it renders RequestDocuments correctly on state NameSubmitted 1`] = `
<div>
  <div
    data-testid="changeofname-RequestDocuments"
  >
    <h3
      class="lbh-heading-h3"
    >
      Supporting documents
    </h3>
    <p
      class="lbh-body-s"
    >
      The tenant needs to provide documents as proof to support their application:
    </p>
    <ul
      class="lbh-list lbh-list--bullet"
    >
      <li>
        <p
          class="lbh-body-s"
          style="font-weight: bold;"
        >
          Two forms of proof of identity. At least one must be photographic ID. Both must contain the tenants new name
        </p>
        <span
          class="lbh-body-s"
        >
          (
          for example: valid passport, driving licence, bank statement, utility bill
          )
        </span>
      </li>
      <li>
        <div>
          <p
            class="lbh-body-s"
            style="font-weight: bold;"
          >
            One of the following documents containing the tenantâ€™s new name:
          </p>
          <ol
            class="lbh-body-s"
            style="margin-left: 2em; margin-top: 15px; font-weight: bold;"
            type="1"
          >
            <li
              style="margin-top: 0px;"
            >
              Marriage certificate
            </li>
            <li
              style="margin-top: 0px;"
            >
              Civil partnership certificate
            </li>
            <li
              style="margin-top: 0px;"
            >
              Decree absolute
            </li>
            <li
              style="margin-top: 0px;"
            >
              Final order
            </li>
            <li
              style="margin-top: 0px;"
            >
              Deed poll document
            </li>
            <li
              style="margin-top: 0px;"
            >
              Statutory declaration
            </li>
          </ol>
        </div>
      </li>
    </ul>
    <h3
      class="lbh-heading-h3"
    >
      Checking supporting documents
    </h3>
    <p
      class="lbh-body-s"
    >
      You can request supporting documents through the Document Evidence Store or you can make an appointment with the tenant to check supporting documents in-person.
    </p>
    <p
      class="lbh-body-s"
    >
      Haylee Macejkovic
       contact details:
    </p>
    <p
      class="lbh-body-s"
    >
      Phone:
      <span
        style="margin-left: 1em;"
      >
        075949281
      </span>
    </p>
    <p
      class="lbh-body-s"
    >
      Email:
      <span
        style="margin-left: 1em;"
      >
        Gay.Jacobi@hotmail.com
      </span>
    </p>
    <form
      action="#"
      class="request-documents-form"
      id="request-documents-form"
      novalidate=""
    >
      <div
        class="govuk-radios lbh-radios"
      >
        <div
          class="govuk-radios__item"
        >
          <input
            class="govuk-radios__input"
            id="requestType-automatic"
            name="requestType"
            type="radio"
            value="automatic"
          />
          <label
            class="govuk-label govuk-radios__label"
            for="requestType-automatic"
          >
            Request documents electronically
          </label>
        </div>
        <div
          class="govuk-radios__item"
        >
          <input
            class="govuk-radios__input"
            id="requestType-manual"
            name="requestType"
            type="radio"
            value="manual"
          />
          <label
            class="govuk-label govuk-radios__label"
            for="requestType-manual"
          >
            I have made an appointment to check supporting documents
          </label>
        </div>
      </div>
      <button
        class="govuk-button lbh-button"
        disabled=""
        type="submit"
      >
        Next
      </button>
    </form>
  </div>
</div>
`;
